google-cloud-pubsublite/src/test/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatchingPublisherTest.java

```diff
@@ -0,0 +1,229 @@
+
+    when(mockRoutingPolicy.routeWithoutKey()).thenReturn(Partition.of(4));
+    publisher.publish(message);
+
+    ApiExceptionMatcher.assertThrowableMatches(
+        publisher.failureCause(), StatusCode.Code.FAILED_PRECONDITION);
+    assertThrows(IllegalStateException.class, publisher::flush);
+    assertThrows(IllegalStateException.class, () -> publisher.publish(Message.builder().build()));
+  }
+
+  @Test
+  public void testChildPublisherFailure() throws Exception {
+    publisher0.fail(new CheckedApiException(StatusCode.Code.FAILED_PRECONDITION));
+
+    ApiExceptionMatcher.assertThrowableMatches(
+        publisher.failureCause(), StatusCode.Code.FAILED_PRECONDITION);
+    assertThrows(IllegalStateException.class, publisher::flush);
+    assertThrows(IllegalStateException.class, () -> publisher.publish(Message.builder().build()));
+  }
+
+  @Test
+  public void testFlush() throws Exception {
+    publisher.flush();
+    verify(publisher0).flush();
+    verify(publisher1).flush();
+  }
+
+  @Test
+  public void testIncreaseSucceeds() throws Exception {
+    leakedConsumer.accept(3L);
+    verify(mockRoutingPolicyFactory).newPolicy(3L);
+
+    Message message0 = Message.builder().setKey(ByteString.copyFromUtf8("0")).build();
+    Message message1 = Message.builder().setKey(ByteString.copyFromUtf8("1")).build();
+    Message message2 = Message.builder().setKey(ByteString.copyFromUtf8("2")).build();
+    when(mockRoutingPolicy.route(message0.key())).thenReturn(Partition.of(0));
+    when(mockRoutingPolicy.route(message1.key())).thenReturn(Partition.of(1));
+    when(mockRoutingPolicy.route(message2.key())).thenReturn(Partition.of(2));
+
+    publisher.publish(message0);
+    publisher.publish(message1);
+    publisher.publish(message2);
+
+    verify(publisher0).publish(message0);
+    verify(publisher1).publish(message1);
+    verify(publisher2).publish(message2);
+  }
+
+  @Test
+  public void testDecreaseFailsPublisher() {
+    leakedConsumer.accept(1L);
+    ApiExceptionMatcher.assertThrowableMatches(
+        publisher.failureCause(), StatusCode.Code.FAILED_PRECONDITION);
+    assertThrows(IllegalStateException.class, publisher::flush);
+    assertThrows(IllegalStateException.class, () -> publisher.publish(Message.builder().build()));
+  }
+
+  @Test
+  public void testNoopConfigUpdate() throws Exception {
+    leakedConsumer.accept(2L);
+    verifyNoMoreInteractions(mockRoutingPolicyFactory);
+
+    Message message0 = Message.builder().setKey(ByteString.copyFromUtf8("0")).build();
+    Message message1 = Message.builder().setKey(ByteString.copyFromUtf8("1")).build();
+    when(mockRoutingPolicy.route(message0.key())).thenReturn(Partition.of(0));
+    when(mockRoutingPolicy.route(message1.key())).thenReturn(Partition.of(1));
+
+    publisher.publish(message0);
+    publisher.publish(message1);
+
+    verify(publisher0).publish(message0);
+    verify(publisher1).publish(message1);
+  }
+
+  @Test
+  public void testStopWorksProperly() {
```

Comment:
Can you add one for stop after increase?

Reply:
Added